include/wayland/mir/wayland/wayland_base.h

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright © 2019 Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Authored By: William Wold <william.wold@canonical.com>
+ */
+
+#ifndef MIR_WAYLAND_OBJECT_H_
+#define MIR_WAYLAND_OBJECT_H_
+
+struct wl_resource;
+struct wl_global;
+struct wl_client;
+
+namespace mir
+{
+namespace wayland
+{
+
+class Resource
```

Comment:
Note that `Resource` and `Global` are part of a published API and ABI that will be implemented by client code.

This means we cannot add data members or virtual member functions without breaking ABI. Are we confident that this will be a sufficiently rare occurrence to be supportable?

Reply:
In the immediate future: no

Hopefully once we get the life-cycle situation sorted out it will be stable enough

Reply:
> Hopefully we can do that before next release.

This rework is the only thing stopping me rolling out the release. 8^)

Reply:
As said in IRC, we should probably just release while this is still unstable, and get it good by next release.

Reply:
Having asked around, we'll do as you suggest. But I'd like to leave this header out of the release. So to make things easy I'll mark this "No Merge" until I've branched,